pkg/storage/stores/tsdb/head_manager.go

```diff
@@ -257,17 +257,61 @@ func (m *HeadManager) Start() error {
 		}
 	}
 
-	walsByPeriod, err := walsByPeriod(m.dir, m.period)
-	if err != nil {
-		return err
+	// build WALs from parent dir if any, these would've been generated before the TSDB multi-store support was added.
+	// TSDB files built in this step would be migrated to the period specific dir on starting the tsdbManager.
+	if err := m.buildLegacyWALs(); err != nil {
+		return errors.Wrap(err, "building tsdb from legacy WAL files")
 	}
-	level.Info(m.log).Log("msg", "loaded wals by period", "groups", len(walsByPeriod))
 
 	// Load the shipper with any previously built TSDBs
 	if err := m.tsdbManager.Start(); err != nil {
 		return errors.Wrap(err, "failed to start tsdb manager")
 	}
 
+	// build tsdb from old WAL files
+	if err := m.buildWALs(m.dir, false); err != nil {
+		return errors.Wrap(err, "building tsdb from old WAL files")
+	}
+
+	err := m.Rotate(time.Now())
+	if err != nil {
+		return errors.Wrap(err, "rotating tsdb head")
+	}
+
+	m.wg.Add(1)
+	go m.loop()
+
+	return nil
+}
+
+func (m *HeadManager) buildLegacyWALs() error {
+	parentDir := filepath.Dir(filepath.Clean(m.dir))
+	if _, err := os.Stat(managerWalDir(parentDir)); err != nil {
+		if os.IsNotExist(err) {
+			// nothing to build
+			return nil
+		}
+
+		return err
+	}
+
+	// ensure scratch and multitenant dir are present as they are used by tsdbManager
+	for _, d := range managerRequiredDirs(parentDir) {
+		if err := util.EnsureDirectory(d); err != nil {
+			return errors.Wrapf(err, "ensuring required legacy directory exists: %s", d)
+		}
+	}
+
+	return m.buildWALs(parentDir, true)
+}
+
+func (m *HeadManager) buildWALs(dir string, legacy bool) error {
```

Comment:
lets name it a bit better to make it clearer
```suggestion
func (m *HeadManager) buildTSDBFromWALs(dir string, legacy bool) error {
```